README.md
# Regulome DB

## Setting Up Your Environment

These are the primary software versions used in production, and you should be able to use them locally:
- Python 3.4.3
- Node 10
- Ubuntu 14.04

### **0. Xcode for Mac OS build tools**  
- Install [Xcode](https://developer.apple.com/xcode/) from the website or Mac App store because the local build will use some of Xcode's compilation tools.
- Install the Xcode commandline tools (the commandline tools may need to be reinstalled after OS and Xcode updates)
  - `xcode-select --install` 

>:star: _Note_: You will need to open Xcode to accept the end-user agreement from the application, or from the commandline run:_  
>- `sudo xcodebuild -license accept`


### **1. Homebrew for Mac OS package management**  
- Verify that [Homebrew](https://brew.sh/) is installed and working properly:  
  - `brew doctor`


### **2. (Mac) Install or update other dependencies:**

```bash
brew install libevent libmagic libxml2 libxslt openssl graphviz nginx
brew install freetype libjpeg libtiff littlecms webp # chromedriver temporarily broken
pip3 install typing
```
>:star: _Note_: This additional step is required for new macOS Sierra installations
>- `brew cask install Caskroom/cask/xquartz`

>:star: _Note_: **Node version mangement with nvm**: If you need to easily switch between **node** versions you may wish to use [nvm](https://github.com/creationix/nvm) instead (not required for most users)
>- `npm install -g nvm`
>- `nvm install 10`
>- `nvm use 10`
>
>:star: _Note_: **Node version mangement with homebrew**: Switching node versions with homebrew (which is not package manager but not a typical version manager) is possible but is not as flexbible for this purpose, e.g.:
>- `brew install node@7`
>- `brew unlink node@6 && brew link --overwrite --force node@7`
>- `node --version`
>- `brew unlink node@7 && brew link --overwrite --force node@6`
>- `node --version`

>:warning: _Note_: The current version of chromedriver doesn’t allow our BDD tests to pass, so you must install the earlier 2.33 version.
>
> Visit <https://chromedriver.storage.googleapis.com/index.html?path=2.33/> and download chromedriver_mac64.zip to your Downloads folder. Then move the file to /usr/local/bin. Verify you have the correct version installed by then entering:
>
>- `chromedriver --version`
>
> It should show, among other things, that you’re running chromedriver 2.33.

>:warning: _Note_: If you need to update Python dependencies (do not do this randomly as you may lose important brew versions of packages you need):
>- `rm -rf encoded/eggs` (then re-run buildout below)
> and possibly
>- `brew update`
>- `brew upgrade`


### **3. Python**  
Regulome requires a UNIX based system (Mac or Linux) and **Python 3.4.3** (but works with 3.5.x):

 - For local development on a Mac, follow the steps below.  For Linux use apt-get or yum as your Linux flavor demands.  You can consult cloud-config.yml for other steps.

- _Note_: Production is currently using the versions above thus your primary development should always work on that version, and you should test that your code works on versions that will be used in production.

- Linux: apt-get install python3.4-dev or equivalent
    
**Mac OSX Python install instructions**  

The Python version management tool `pyenv` is very useful. 

>:warning: _Note: If you have previously installed python3 from homebrew, you may possibly wish to uninstall it (not required):_  
> - `brew uninstall --force python3`


    
**Install `pyenv` and set the default versions:**
```bash 
brew install pyenv
pyenv install 3.4.3
pyenv install 2.7.13
echo 'export PYENV_ROOT="$HOME/.pyenv"' >> ~/.bash_profile
echo 'export PATH="$PYENV_ROOT/bin:$PATH"' >> ~/.bash_profile
echo 'eval "$(pyenv init -)"' >> ~/.bash_profile
echo 'eval "pyenv shell 2.7.13 3.4.3"' >> ~/.bash_profile
source ~/.bash_profile
```

>:star: _Note: Migrating `pyenv` Python packages_  
>
>_If you have previously installed a Python version from `pyenv`, and want to quickly migrate all your pypi packages to a new version (Python 2 to 2, and Python 3 to 3 only):_
>  - `brew install pyenv-pip-migrate`  
>    
>Example if you previously installed `2.7` which really is _`2.7.0`_:  
>  - `pyenv install 2.7.13`  
>  - `pyenv migrate 2.7 2.7.13`

>:star: _Note: `pyenv` install fails with "ERROR: The Python ssl extension was not compiled. Missing the OpenSSL lib?" for MAC OS High Sierra
>
>Uninstall and re-install openssl using the following command when you install pyenv
>- `brew uninstall --ignore-dependencies openssl && brew install openssl && CFLAGS="-I$(brew --prefix openssl)/include" LDFLAGS="-L$(brew --prefix openssl)/lib" pyenv install <VERSION>`

**Set the correct Python for the current directory:**
```bash
pyenv local 3.4.3
```

### **4. Run buildout:**

- `pip3 install -U zc.buildout setuptools`
- `pyenv rehash`
- `buildout bootstrap`
- `bin/buildout`

- Edit the key `genomic_data_service_url` on `development.ini` with the correct Genomic Data Services URL.

>:star: _Note_: If you have issues with Pillow you may need to install new xcode command line tools:
> Update or Install [Xcode](https://developer.apple.com/xcode/) from the Mac AppStore (reboot may be required) and re-run:  
> - `xcode-select --install`  


>:star: _Note_: **Clean ALL the Things!** If you wish to **completely rebuild** the application or **cleanly reload** dependencies (:warning: long re-build time!):  
>- `make clean && buildout bootstrap && bin/buildout`


### **5. Start the application locally**


- **Terminal window**:
  In a terminal, run the app with:

  - `bin/pserve development.ini`


### **6. :tada: Check out the app! :tada:**
- Browse to the interface at http://localhost:8000/.


## Running tests

- To run specific tests locally:
    
  `bin/test -k test_name`
    
- To run with a debugger:
    
  `bin/test --pdb`

- Specific tests to run locally for schema changes:

  `bin/test -k test_load_workbook`  
  `bin/test -k test_load_schema`

- Run the Pyramid tests with:

  `bin/test -m "not bdd"`

- Run the Browser tests with:

  `bin/test -m bdd -v --splinter-webdriver chrome`

- Run the Javascript tests with:

  `npm test`

- Or if you need to supply command line arguments:

  `./node_modules/.bin/jest`

- **Test ALL the things!**
  
  `bin/test -v -v --splinter-webdriver chrome && npm test`


## Building Javascript and CSS


Our Javascript is written using ES6 and JSX, so needs to be compiled using babel and webpack. Our CSS is written in the SCSS variant of [Sass](http://sass-lang.com/) and also needs compilation using webpack.

- To re-build **production-ready** bundles, do:

  `npm run build`

  (This is also done as part of running buildout.)

- To build **development** bundles and continue updating them as you edit source files, run:

  `npm run dev`

The development bundles are not minified, to speed up building. The above command runs continually in your terminal window and watches for changes in Javascript and SCSS files, rebuilding the bundles as you make changes.

## Creating a demo machine


- After buildout you (if you have the correct permissions) can run for a single-node "cluster":

  `bin/deploy`
  
  
- The script above will spin up a server in the AWS cloud with the current branch, and with a computed nameserver alias based on the branch and your username.  There are options to use a different branch and/or different instance name and also if you want to use AWS spot instances...and you can specify which AWS profile you want to use.   

  
- Deploy script help (how to specify name, instance size, etc):

  `bin/deploy --help`
      
For all options, including setting up ES clusters (needed for full production).  After indexing (currently 8+hrs) the machine can be downsized at AWS to an m4.2xlarge, unless you are planning to submit significant data to it.


## Linting your code within your code editor

To set up linting with [Sublime Text 3](https://www.sublimetext.com/3) or [Visual Studio Code](https://code.visualstudio.com/), first install the linters:

```bash
pip3 install flake8
npm install -g eslint
npm install -g eslint-plugin-react
```
>:warning: _Note_: You don't have to use Sublime Text 3 but you *must* insure that linting in your editor behaves as it does in Sublime Text 3.



### Sublime Text 3

After first setting up [Package Control](https://packagecontrol.io/) (follow install and usage instructions on site), use it to install the following linting packages in Sublime Text 3:

 - `sublimelinter`
 - `sublimelinter-flake8`
 - `SublimeLinter-contrib-eslint` ([Sublime linter eslint instructions](https://github.com/roadhump/SublimeLinter-eslint#plugin-installation))
 - `babel` ([Babel instructions](https://github.com/babel/babel-sublime#setting-as-the-default-syntax))

**Sublime Linting with `pyenv`**   
To get Sublime to lint Python code using `pyenv` you must add the python version and paths and python_paths to your Sublime Linter Preferences. In **Sublime Text**, navigate to the user linter preferences:  

- Sublime Preferences  -> Package Settings -> Sublime Linter -> Settings-User

- Add the following (modify existing preference settings file or add this entire JSON object below if the file is blank):

```json
{
    "user": {
        "@python": 3.4,
        
        "paths": {
            "linux": [],
            "osx": [
                "/Users/YOURUSERNAME/.pyenv/versions/3.4.3/bin/",
                "/Users/YOURUSERNAME/.pyenv/versions/2.7/bin/"
            ],
            "windows": []
        },
        "python_paths": {
            "linux": [],
            "osx": [
                "/Users/YOURUSERNAME/.pyenv/versions/3.4.3/bin/python3",

CHANGES.rst
Changes
=======

2.0.3 Release  (Data Services Release)
--------------------------------
Footprinting files were adjusted to correct assembly
REG-227 update experiment annotation links
REG-198-add-download-button
REG-229 Convert backend to Dataservices
	1.	dataset key fix
	2.	adding file download redirect for genome browse
	3.	redirecting to coordinates page when summary has one result
	4.	adding encode redirect for cors requests
	5.	avoiding 301 to www and losing CORS headers
	6.	using Encode as source of files
	7.	fixing absolute hrefs for accessibility widget
	8.	fixing absolute hrefs for widgets
	9.	valis bug fix, checking if facet exists before 
	10.	adding absolute hrefs for search
	11.	fix relative href
	12.	making data service url configurable
	13.	removing postgres and ES references
	14.	adding custom 404 page
	15.	clean up unused methods
	16.	using gp3 for EBS
	17.	removing ES and Postgres from deployment
	18.	restoring deploy script
	19.	cleaning up buildout and configs
	20.	removing indexer endpoints
	21.	removing need of DB and/or ES
	22.	removing profiles-title fetch
	23.	removing snovault, sign-on, help pages from db
	24.	turning help page into a static page
	25.	removing test workbook
	26.	removing cart and vis indexer
	27.	removing configs
	28.	handling file downloads
	29.	connecting search endpoints with data service


2.0.2 release - virtual (snovault 1.0.26)
--------------------------------
REG-217-default-MAF-filter to 0
REG-209-script-search-snps-in-region (#104)
REG-212-update-IC_max-precalculated-bigWig (#105)
REG-210-fix-download-summary-bug (#103)
always 16GB heap for ES
revert s3 uri for href
using s3_uri as href for files when available
removing proxy servers for downloads
updating regulome s3 paths
REG-200 Add Elasticsearch log compression cron job (#95)
REG-199-index-filter-MAF (#93)
REG-197-change-homepage-title (#92)
REG-196-parallelize-regulome-search-script (#91)
REG-194-trim-script-output-for-peak-information (#90)
REG-191-Change-JVM-GC-to-G1GC (#88)


2.0.1 release - virtual (snovault 1.0.26)
--------------------------------
remove beta banner (#84)
fix google analytics (#87)
REG-186-restrict-to-old-chromatin-state-data (#86)
REG-187-optimize-script-memory-usage (#85)
add link to legacy site (#82)
accessibility page redesign (#75)
REG-161-add-marked-location (#76)
REG-169 return matched PWM peaks in script (#74)
REG-14 update dbSNP reference (#78)
REG-175 update coordinates examples (#73)
REG-178 fix motif error (#77)
REG-174 Freeze setuptools at v43 (#72)
REG-171 add reset for browser (#70)
REG-142 add regulome summary download (#69)
REG-170 fix accessibility page alignment (#67)
REG-165 update motif summaries (#64)
REG-152 scoring script
REG-127 Ignore archived files if there is any released file


2.0.0 BETA release virtual (snovault 1.0.26)
--------------------------------
REG-68 Use all ENCODE data
REG-168 fix ENCODE file download (#63)
REG-164 update motif colors (#62)
REG-137 add filters to browser (#57)
REG-162 collapse and align motifs (#58)
REG-148 regulome-search to process one arbitrary region (#60)
REG-163 fix discrepancy in MAF cutoff (#59)
REG-158 update regulome tests for scores (#56)
REG-143 fix browser issues (#47)
REG-155 sort table by score properly (#55)
REG-154 remove ChIP and DNase signals from scoring algorithm (#53)
REG-147 upgrade React (#51)
REG-153 add beta banner (#52)
REG-151 fix chromatin graphic (#54)
REG-119 motif alignments (#50)
REG-136 add back button functionality (#46)
REG-145 updates to beta website (#48)
REG-149 fix table sorting (#49)
REG-121 update QTL thumbnail (#45)
REG-144 fix accessibility counts (#44)
REG-140 split summary table (#43)
REG-116 Genome browser (#42)
REG-91 chromatin visualization (#41)
REG-128 use peak metadata (#40)
REG-130 fix query (#39)
REG-74 Summary graphic and clickable thumbnails (#38)
REG-124 update scoring algorithm (#37)
REG-120 index value and strand (#32)
REG-126-update-to-java-install-manual (#36)
REG-123 Get feature values from pre-calculated bigWig files (#31)
REG-104 make search result smaller (#30)
REG-106 fix for biosample in details plus tests (#25)
REG-110 Downgrade scikit-learn (#28)
REG-108 REG-111 mixed tests (#29)
REG-109 update sauce connect to 4.5.3 (#27)
REG-105 manual rebase (#24)
REG-101 invert index (#22)
REG-71 add PWM logos (#21)
REG-97 split regulome (#20)
REG-102 remove cart test (#18)
REG-94 IE bug fixes (#14)
REG-84 new regulome scoring algorithm (#7)
ENCD-4581 Temp install java 8 through aws
REG-90 Encode-related updates
SNO-73 Add uuid queue module
ENCD-4456 Override update objs in vis reg indexers
ENCD-4240 Allow review characterizations (#2591)
REG-92 fix clickable FAQ for Safari and Firefox (#13)
REG-83 fix overlapping icons (#11)
REG-88 fix biosample info (#9)
REG-89 update logos (#10) 

src/encoded/loadxl.py
from past.builtins import basestring
from .typedsheets import cast_row_values
from functools import reduce
import io
import logging
import os.path

text = type(u'')

DEFAULT_LOG_LEVEL = logging.INFO
logger = logging.getLogger('encoded')
logger.setLevel(DEFAULT_LOG_LEVEL)  # doesn't work to shut off sqla INFO


ORDER = [
    'user',
    'award',
    'lab',
    'organism',
    'source',
    'gene',
    'target',
    'publication',
    'document',
    'antibody_lot',
    'biosample_type',
    'antibody_characterization',
    'treatment',
    'genetic_modification',
    'genetic_modification_characterization',
    'mouse_donor',
    'fly_donor',
    'worm_donor',
    'human_donor',
    'donor_characterization',
    'biosample',
    'biosample_characterization',
    'platform',
    'library',
    'experiment',
    'replicate',
    'annotation',
    'project',
    'publication_data',
    'reference',
    'ucsc_browser_composite',
    'matched_set',
    'treatment_time_series',
    'treatment_concentration_series',
    'organism_development_series',
    'replication_timing_series',
    'aggregate_series',
    'reference_epigenome',
    'software',
    'software_version',
    'analysis_step',
    'analysis_step_version',
    'pipeline',
    'analysis_step_run',
    'file',
    'star_quality_metric',
    'bismark_quality_metric',
    'cpg_correlation_quality_metric',
    'chipseq_filter_quality_metric',
    'samtools_flagstats_quality_metric',
    'mad_quality_metric',
    'correlation_quality_metric',
    'edwbamstats_quality_metric',
    'hotspot_quality_metric',
    'idr_summary_quality_metric',
    'complexity_xcorr_quality_metric',
    'duplicates_quality_metric',
    'filtering_quality_metric',
    'trimming_quality_metric',
    'samtools_stats_quality_metric',
    'idr_quality_metric',
    'histone_chipseq_quality_metric',
    'generic_quality_metric',
    'image',
    'page',
    'cart'
]

IS_ATTACHMENT = [
    'attachment',
    'IDR_plot_true',
    'IDR_plot_rep1_pr',
    'IDR_plot_rep2_pr',
    'IDR_plot_pool_pr',
    'IDR_parameters_true',
    'IDR_parameters_rep1_pr',
    'IDR_parameters_rep2_pr',
    'IDR_parameters_pool_pr',
    'cross_correlation_plot'
]


def _reset_log_level(log_level):
    '''
    Resets the default log level, usually for running bin tests
    '''
    if logger and log_level is not DEFAULT_LOG_LEVEL:
        numeric_level = getattr(logging, log_level.upper(), None)
        if not isinstance(numeric_level, int):
            raise ValueError('Invalid log level: %s' % log_level)
        logger.setLevel(log_level)


##############################################################################
# Pipeline components
#
# http://www.stylight.com/Numbers/pipes-and-filters-architectures-with-python-generators/
#
# Dictionaries are passed through the pipeline. By convention, values starting
# with an underscore (_) are ignored by the component posting a final value
# so are free for communicating information down the pipeline.


def noop(dictrows):
    """ No-op component

    Useful for pipeline component factories.
    """
    return dictrows


def remove_keys_with_empty_value(dictrows):
    for row in dictrows:
        yield {
            k: v for k, v in row.items()
            if k and v not in ('', None, [])
        }


##############################################################################
# Pipeline component factories


def warn_keys_with_unknown_value_except_for(*keys):
    def component(dictrows):
        for row in dictrows:
            for k, v in row.items():
                if k not in keys and text(v).lower() == 'unknown':
                    logger.warn('unknown %r for %s' % (k, row.get('uuid', '<empty uuid>')))
            yield row

    return component


def skip_rows_missing_all_keys(*keys):
    def component(dictrows):
        for row in dictrows:
            if not any(key in row for key in keys):
                row['_skip'] = True
            yield row

    return component


def skip_rows_with_all_key_value(**kw):
    def component(dictrows):
        for row in dictrows:
            if all(row[k] == v if k in row else False for k, v in kw.items()):
                row['_skip'] = True
            yield row

    return component


def skip_rows_without_all_key_value(**kw):
    def component(dictrows):
        for row in dictrows:
            if not all(row[k] == v if k in row else False for k, v in kw.items()):
                row['_skip'] = True
            yield row

    return component


def remove_keys(*keys):
    def component(dictrows):
        for row in dictrows:
            for key in keys:
                row.pop(key, None)
            yield row

    return component


def skip_rows_with_all_falsey_value(*keys):
    def component(dictrows):
        for row in dictrows:
            if all(not row[key] if key in row else False for key in keys):
                row['_skip'] = True
            yield row

    return component


def add_attachments(docsdir):
    def component(dictrows):
        for row in dictrows:
            for attachment_property in IS_ATTACHMENT:
                filename = row.get(attachment_property, None)
                if filename is None:
                    continue
                try:
                    path = find_doc(docsdir, filename)
                    row[attachment_property] = attachment(path)
                except ValueError as e:
                    row['_errors'] = repr(e)
            yield row

    return component


##############################################################################
# Read input from spreadsheets
#
# Downloading a zipfile of xlsx files from Google Drive is most convenient
# but it's better to check tsv into git.


def read_single_sheet(path, name=None):
    """ Read an xlsx, csv or tsv from a zipfile or directory
    """
    from zipfile import ZipFile
    from . import xlreader

    if name is None:
        root, ext = os.path.splitext(path)
        stream = open(path, 'r')

        if ext == '.xlsx':
            return read_xl(stream)

        if ext == '.tsv':
            return read_csv(stream, dialect='excel-tab')

        if ext == '.csv':
            return read_csv(stream)

        if ext == '.json':
            return read_json(stream)

        raise ValueError('Unknown file extension for %r' % path)

    if path.endswith('.xlsx'):
        return cast_row_values(xlreader.DictReader(open(path, 'rb'), sheetname=name))

    if path.endswith('.zip'):
        zf = ZipFile(path)
        names = zf.namelist()

        if (name + '.xlsx') in names:
            stream = zf.open(name + '.xlsx', 'r')
            return read_xl(stream)

        if (name + '.tsv') in names:
            stream = io.TextIOWrapper(zf.open(name + '.tsv'), encoding='utf-8')
            return read_csv(stream, dialect='excel-tab')

        if (name + '.csv') in names:
            stream = io.TextIOWrapper(zf.open(name + '.csv'), encoding='utf-8')
            return read_csv(stream)

        if (name + '.json') in names:
            stream = io.TextIOWrapper(zf.open(name + '.json'), encoding='utf-8')
            return read_json(stream)

    if os.path.isdir(path):
        root = os.path.join(path, name)

        if os.path.exists(root + '.xlsx'):
            stream = open(root + '.xlsx', 'rb')
            return read_xl(stream)

        if os.path.exists(root + '.tsv'):
            stream = open(root + '.tsv', 'rU')
            return read_csv(stream, dialect='excel-tab')

        if os.path.exists(root + '.csv'):
            stream = open(root + '.csv', 'rU')
            return read_csv(stream)

        if os.path.exists(root + '.json'):
            stream = open(root + '.json', 'r')
            return read_json(stream)

    return []


def read_xl(stream):
    from . import xlreader
    return cast_row_values(xlreader.DictReader(stream))


def read_csv(stream, **kw):
    import csv
    return cast_row_values(csv.DictReader(stream, **kw))


def read_json(stream):
    import json
    obj = json.load(stream)
    if isinstance(obj, dict):
        return [obj]
    return obj


##############################################################################
# Posting json
#
# This would a one liner except for logging

def request_url(item_type, method):
    def component(rows):
        for row in rows:
            if method == 'POST':
                url = row['_url'] = '/' + item_type
                yield row
                continue

            if '@id' in row:
                url = row['@id']
                if not url.startswith('/'):
                    url = '/' + url
                row['_url'] = url
                yield row
                continue

            # XXX support for aliases
            for key in ['uuid', 'accession']:
                if key in row:
                    url = row['_url'] = '/' + row[key]
                    break
            else:
                row['_errors'] = ValueError('No key found. Need uuid or accession.')

            yield row

    return component


def make_request(testapp, item_type, method):
    json_method = getattr(testapp, method.lower() + '_json')

    def component(rows):
        for row in rows:
            if row.get('_skip') or row.get('_errors') or not row.get('_url'):
                continue

            # Keys with leading underscores are for communicating between
            # sections
            value = row['_value'] = {
                k: v for k, v in row.items() if not k.startswith('_') and not k.startswith('@')
            }

            url = row['_url']
            row['_response'] = json_method(url, value, status='*')

            yield row

    return component


##############################################################################
# Logging


def trim(value):
    """ Shorten excessively long fields in error log
    """
    if isinstance(value, dict):
        return {k: trim(v) for k, v in value.items()}
    if isinstance(value, list):
        return [trim(v) for v in value]
    if isinstance(value, basestring) and len(value) > 160:
        return value[:77] + '...' + value[-80:]
    return value


def pipeline_logger(item_type, phase):
    def component(rows):
        created = 0
        updated = 0
        errors = 0
        skipped = 0
        count = 0
        for index, row in enumerate(rows):
            row_number = index + 2  # header row
            count = index + 1
            res = row.get('_response')

            if res is None:
                _skip = row.get('_skip')
                _errors = row.get('_errors')
                if row.get('_skip'):
                    skipped += 1
                elif _errors:
                    errors += 1
                    logger.error('%s row %d: Error PROCESSING: %s\n%r\n' % (item_type, row_number, _errors, trim(row)))
                yield row
                continue

            url = row.get('_url')
            uuid = row.get('uuid')

            if res.status_int == 200:
                updated += 1
                logger.debug('UPDATED: %s' % url)

            if res.status_int == 201:
                created += 1
                logger.debug('CREATED: %s' % res.location)

            if res.status_int == 409:
                logger.error('CONFLICT: %r' % res.json['detail'])

            if res.status_int == 422:
                logger.error('VALIDATION FAILED: %r' % trim(res.json['errors']))

            if res.status_int // 100 == 4:
                errors += 1
                logger.error('%s row %d: %s (%s)\n%r\n' % (item_type, row_number, res.status, url, trim(row['_value'])))

            yield row

        loaded = created + updated
        logger.info('Loaded %d of %d %s (phase %s). CREATED: %d, UPDATED: %d, SKIPPED: %d, ERRORS: %d' % (
            loaded, count, item_type, phase, created, updated, skipped, errors))

    return component


##############################################################################
# Attachments


def find_doc(docsdir, filename):
    path = None
    for dirpath in docsdir:
        candidate = os.path.join(dirpath, filename)
        if not os.path.exists(candidate):
            continue
        if path is not None:
            msg = 'Duplicate filenames: %s, %s' % (path, candidate)
            raise ValueError(msg)
        path = candidate
    if path is None:
        raise ValueError('File not found: %s' % filename)
    return path


def attachment(path):
    """ Create an attachment upload object from a filename

    Embeds the attachment as a data url.
    """
    import magic
    import mimetypes
    from PIL import Image
    from base64 import b64encode

    filename = os.path.basename(path)
    mime_type, encoding = mimetypes.guess_type(path)
    major, minor = mime_type.split('/')
    detected_type = magic.from_file(path, mime=True).decode('ascii')

    # XXX This validation logic should move server-side.
    if not (detected_type == mime_type or
            detected_type == 'text/plain' and major == 'text'):
        raise ValueError('Wrong extension for %s: %s' % (detected_type, filename))

    with open(path, 'rb') as stream:
        attach = {
            'download': filename,
            'type': mime_type,
            'href': 'data:%s;base64,%s' % (mime_type, b64encode(stream.read()).decode('ascii'))
        }

        if mime_type in ('application/pdf', 'text/plain', 'text/tab-separated-values', 'text/html'):
            # XXX Should use chardet to detect charset for text files here.
            return attach

        if major == 'image' and minor in ('png', 'jpeg', 'gif', 'tiff'):
            # XXX we should just convert our tiffs to pngs
            stream.seek(0, 0)
            im = Image.open(stream)
            im.verify()
            if im.format != minor.upper():
                msg = "Image file format %r does not match extension for %s"
                raise ValueError(msg % (im.format, filename))

            attach['width'], attach['height'] = im.size
            return attach

    raise ValueError("Unknown file type for %s" % filename)


##############################################################################
# Pipelines


def combine(source, pipeline):
    """ Construct a combined generator from a source and pipeline
    """
    return reduce(lambda x, y: y(x), pipeline, source)


def process(rows):
    """ Pull rows through the pipeline
    """
    for row in rows:
        pass


def get_pipeline(testapp, docsdir, test_only, item_type, phase=None, method=None):
    pipeline = [
        skip_rows_with_all_key_value(test='skip'),
        skip_rows_with_all_key_value(_test='skip'),
        skip_rows_with_all_falsey_value('test') if test_only else noop,
        skip_rows_with_all_falsey_value('_test') if test_only else noop,
        remove_keys_with_empty_value,
        skip_rows_missing_all_keys('uuid', 'accession', '@id', 'name'),
        remove_keys('schema_version'),
        warn_keys_with_unknown_value_except_for(
            'lot_id', 'sex', 'life_stage', 'health_status', 'ethnicity',
            'strain_background', 'age', 'version',
            'model_organism_health_status',
            'model_organism_age',
            'model_organism_sex',
            'mouse_life_stage',
            # 'flowcell_details.machine',
        ),
        add_attachments(docsdir),
    ]
    if phase == 1:
        method = 'POST'
        pipeline.extend(PHASE1_PIPELINES.get(item_type, []))
    elif phase == 2:
        method = 'PUT'
        pipeline.extend(PHASE2_PIPELINES.get(item_type, []))

    pipeline.extend([
        request_url(item_type, method),
        remove_keys('uuid') if method in ('PUT', 'PATCH') else noop,
        make_request(testapp, item_type, method),
        pipeline_logger(item_type, phase),
    ])
    return pipeline


# Additional pipeline sections for item types

PHASE1_PIPELINES = {
    'user': [
        remove_keys('lab', 'submits_for'),
    ],
    'biosample': [
        remove_keys('derived_from', 'pooled_from', 'part_of', 'host'),
    ],
    'library': [
        remove_keys('spikeins_used'),
    ],
    'experiment': [
        remove_keys('possible_controls', 'related_files', 'supersedes'),
    ],
    'mouse_donor': [
        remove_keys('parent_strains'),
    ],
    'fly_donor': [
        remove_keys('parent_strains'),
    ],
    'worm_donor': [
        remove_keys('outcrossed_strain', 'parent_strains'),
    ],
    'human_donor': [
        remove_keys('parents', 'children', 'siblings', 'twin'),
    ],
    'publication': [
        remove_keys('datasets'),
    ],
    'annotation': [
        remove_keys('related_files', 'software_used'),
    ],
    'project': [
        remove_keys('related_files'),
    ],
    'publication_data': [
        remove_keys('related_files'),
    ],
    'reference': [
        remove_keys('related_files', 'software_used'),
    ],
    'ucsc_browser_composite': [
        remove_keys('related_files'),
    ],
    'treatment_time_series': [
        remove_keys('related_datasets'),
    ],
    'treatment_concentration_series': [
        remove_keys('related_datasets'),
    ],
    'aggregate_series': [
        remove_keys('related_datasets'),
    ],
    'organism_development_series': [
        remove_keys('related_datasets'),
    ],
    'replication_timing_series': [
        remove_keys('related_datasets'),
    ],
    'reference_epigenome': [
        remove_keys('related_datasets', 'supersedes'),
    ],
    'matched_set': [
        remove_keys('related_datasets'),
    ],
    'file': [
        remove_keys('derived_from', 'controlled_by', 'supersedes')
    ],
    'analysis_step': [
        remove_keys('parents')
    ],
    'treatment': [
        remove_keys('biosamples_used')
    ]

}


##############################################################################
# Phase 2 pipelines
#
# A second pass is required to cope with reference cycles. Only rows with
# filtered out keys are updated.


PHASE2_PIPELINES = {
    'user': [
        skip_rows_missing_all_keys('lab', 'submits_for'),
    ],
    'biosample': [
        skip_rows_missing_all_keys('derived_from', 'pooled_from', 'part_of', 'host'),
    ],
    'library': [
        skip_rows_missing_all_keys('spikeins_used'),
    ],
    'experiment': [
        skip_rows_missing_all_keys('related_files', 'possible_controls', 'supersedes'),
    ],
    'human_donor': [
        skip_rows_missing_all_keys('parents', 'children ', 'siblings', 'twin'),
    ],
    'mouse_donor': [
        skip_rows_missing_all_keys('parent_strains'),
    ],
    'worm_donor': [
        skip_rows_missing_all_keys('outcrossed_strain', 'parent_strains'),
    ],
    'fly_donor': [
        skip_rows_missing_all_keys('parent_strains'),
    ],
    'annotation': [
        skip_rows_missing_all_keys('related_files', 'software_used'),
    ],
    'project': [
        skip_rows_missing_all_keys('related_files'),
    ],
    'publication_data': [
        skip_rows_missing_all_keys('related_files'),
    ],
    'reference': [
        skip_rows_missing_all_keys('related_files', 'software_used'),
    ],
    'ucsc_browser_composite': [
        skip_rows_missing_all_keys('related_files'),
    ],
    'treatment_time_series': [
        skip_rows_missing_all_keys('related_datasets'),
    ],
    'aggregate_series': [
        skip_rows_missing_all_keys('related_datasets'),
    ],
    'treatment_concentration_series': [
        skip_rows_missing_all_keys('related_datasets'),
    ],
    'organism_development_series': [
        skip_rows_missing_all_keys('related_datasets'),
    ],
    'replication_timing_series': [
        skip_rows_missing_all_keys('related_datasets'),
    ],
    'reference_epigenome': [
        skip_rows_missing_all_keys('related_datasets', 'supersedes'),
    ],
    'matched_set': [
        skip_rows_missing_all_keys('related_datasets'),
    ],
    'publication': [
        skip_rows_missing_all_keys('datasets'),
    ],
    'file': [
        skip_rows_missing_all_keys('derived_from', 'controlled_by', 'supersedes')
    ],
    'analysis_step': [
        skip_rows_missing_all_keys('parents')
    ],
    'treatment': [
        skip_rows_missing_all_keys('biosamples_used')
    ]
}


def load_all(testapp, filename, docsdir, log_level=None, test=False):
    if log_level is not None:
        _reset_log_level(log_level)
    for item_type in ORDER:
        try:
            source = read_single_sheet(filename, item_type)
        except ValueError:
            logger.error('Opening %s %s failed.', filename, item_type)
            continue
        pipeline = get_pipeline(testapp, docsdir, test, item_type, phase=1)
        process(combine(source, pipeline))

    for item_type in ORDER:
        if item_type not in PHASE2_PIPELINES:
            continue
        try:
            source = read_single_sheet(filename, item_type)
        except ValueError:
            continue
        pipeline = get_pipeline(testapp, docsdir, test, item_type, phase=2)
        process(combine(source, pipeline))


def load_test_data(app):
    from webtest import TestApp
    environ = {
        'HTTP_ACCEPT': 'application/json',
        'REMOTE_USER': 'TEST',
    }
    testapp = TestApp(app, environ)

    from pkg_resources import resource_filename
    inserts = resource_filename('encoded', 'tests/data/inserts/')
    docsdir = [resource_filename('encoded', 'tests/data/documents/')]
#    load_all(testapp, inserts, docsdir)
